src/resources/cpp/lambdas.md

```diff
@@ -0,0 +1,208 @@
+
+if they are similar to functions one might ask why would I ever need a lambda? They have multiple usages but the simplest and most common one is for predicates in algorithms. Algorithms are nice and generic and you want their behavior to be slightly different depending on your specific use case.
+
+```cpp
+#include <algorithm>
+#include <iostream>
+#include <vector>
+
+int main()
+{
+    std::vector<int> v{3, -4, 2, -8, 15, 267};
+
+    // Here we declare a local lambda object, called print
+    auto print = [](const int& n) { std::cout << n << '\n'; };
+
+    // Here we use this lambda object in the algorithm for each, where the lambda will be called for each
+    // iteration and print the value in the vector.
+    std::for_each(v.cbegin(), v.cend(), print);
+
+    // As a lambda is similar to a function we can also call it like a function. This will print 5
+    print(5);
+}
+```
+
+## Structure of lambdas
+
+A lambda consists of at least three parts:
+
+- `[]`, the introducer, containing captures.
+- `()`, the parameters (optional in some cases).
+- `{}`, the body of the lambda.
+
+There are many further optional parts, but `[](){}` or `[]{}` are the bare minimum, depending on the language version.
+
+### Parameters `()`
+
+The parameter part of a lambda is the exact same as the parameter list of a function. Here you specify the inputs of your lambda.
+
+For example `[](int a, int b){}` is a lambda that takes in 2 integer values.
+
+### body `{}`
+
+The lambda body is the exact same as the body of the function. It's hold the code that the lambda will execute when it's invoked.
+
+### Capture list
+
+As the name implies, the capture list allows us to capture values into the lambda as it's constructed. Unlike the parameters where we specify which values we want to pass in when we invoke the lambda. This tends to be particularly common when you want to bring external state into a generic algorithm.
+
+Take for example a simple program where you want to multiply every number with a number the user provides.
+
+```cpp
+#include <algorithm>
+#include <iostream>
+#include <vector>
+
+int main()
+{
+    std::vector<int> v{3, -4, 2, -8, 15, 267};
+
+    std::cout << "Please enter a number to multiply with";
+    int multiply_number = 0;
+    std::cin >> multiply_number;
+
+    // Here we declare a local lambda object. We capture the variable multiply_number so it can be used in the lambda. We return the value of the passed in value with multiply_number.
+    auto func = [multiply_number](const int& n) {
+        return n * multiply_number;
+    };
+
+    // Here we use this lambda object in the algorithm transform. This algorithm takes a ranges to iterate over, an output destination (the begin of the same vector in this case) and the predicate (our lambda)
+    std::transform(v.cbegin(), v.cend(), v.begin(), func);
```

Comment:
Explicitly `return 0;`  We haven't introduced (and should not introduce to beginners) the concept of implicit return 0.